File: apppickerview/apppickerview/src/main/java/androidx/apppickerview/widget/AbsAdapter.java
/*
 * Copyright 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.apppickerview.widget;

import static androidx.annotation.RestrictTo.Scope.LIBRARY_GROUP_PREFIX;

import android.content.ComponentName;
import android.content.Context;
import android.icu.text.AlphabeticIndex;
import android.os.Build;
import android.os.LocaleList;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.style.ForegroundColorSpan;
import android.util.Log;
import android.util.TypedValue;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.SectionIndexer;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.annotation.RestrictTo;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import androidx.reflect.text.SeslTextUtilsReflector;

import java.text.Collator;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.StringTokenizer;

/*
 * Original code by Samsung, all rights reserved to the original author.
 */

@RestrictTo(LIBRARY_GROUP_PREFIX)
public abstract class AbsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder>
        implements Filterable, SectionIndexer {
    private static final String TAG = "AppPickerViewAdapter";

    private static final int MAX_OFFSET = 200;

    private final AppPickerIconLoader mAppPickerIconLoader;
    @NonNull
    protected Context mContext;
    private AppPickerView.OnBindListener mOnBindListener;
    AppPickerView.OnSearchFilterListener mOnSearchFilterListener;
    final List<AppPickerView.AppLabelInfo> mDataSet = new ArrayList<>();
    final List<AppPickerView.AppLabelInfo> mDataSetFiltered = new ArrayList<>();
    private final Map<String, Integer> mSectionMap = new HashMap<>();
    private String[] mSections = new String[0];
    String mSearchText = "";

    boolean mHideAllApps = false;

    private final int mForegroundColor;
    private int mOrder;
    private int[] mPositionToSectionIndex;
    protected int mType;

    public AbsAdapter(@NonNull Context context, int type, int order, @Nullable AppPickerIconLoader iconLoader, boolean showDivider) {
        mContext = context;
        mType = type;
        mOrder = order;
        mAppPickerIconLoader = iconLoader;

        TypedValue outValue = new TypedValue();
        mContext.getTheme().resolveAttribute(androidx.appcompat.R.attr.colorPrimary, outValue, true);
        mForegroundColor = outValue.resourceId != 0 ?
                ContextCompat.getColor(mContext, outValue.resourceId) : outValue.data;
    }

    static AbsAdapter getAppPickerAdapter(Context context, List<String> packageNamesList, int type,
            int order, List<AppPickerView.AppLabelInfo> labelInfoList,
            AppPickerIconLoader iconLoader,
            List<ComponentName> activityNamesList,
             boolean showDivider) {
        final AbsAdapter adapter;
        if (type >= AppPickerView.TYPE_GRID) {
            adapter = new GridAdapter(context, type, order, iconLoader, showDivider);
        } else {
            adapter = new ListAdapter(context, type, order, iconLoader, showDivider);
        }
        adapter.setHasStableIds(true);
        adapter.resetPackages(packageNamesList, false, labelInfoList, activityNamesList);
        return adapter;
    }

    List<AppPickerView.AppLabelInfo> getDataSet() {
        return mDataSet;
    }

    void resetPackages(List<String> packageNamesList,
            boolean dataSetchanged,
            List<AppPickerView.AppLabelInfo> labelInfoList,
            List<ComponentName> activityNamesList) {
        Log.i(TAG, "Start resetpackage dataSetchanged : " + dataSetchanged);

        mDataSet.clear();
        mDataSet.addAll(DataManager.resetPackages(mContext, packageNamesList, labelInfoList, activityNamesList));

        if (Build.VERSION.SDK_INT >= 24) {
            if (getAppLabelComparator(mOrder) != null) {
                mDataSet.sort(getAppLabelComparator(mOrder));
            }
        }

        if (hasAllAppsInList()) {
            if (!mDataSet.isEmpty()) {
                mDataSet.add(0,
                        new AppPickerView.AppLabelInfo(AppPickerView.ALL_APPS_STRING,
                                "", ""));
            }
        }
        mDataSetFiltered.clear();
        mDataSetFiltered.addAll(mDataSet);

        refreshSectionMap();

        if (dataSetchanged) {
            notifyDataSetChanged();
        }

        Log.i(TAG, "End resetpackage");
    }

    void addPackage(int position, String label) {
        mDataSet.add(position,
                new AppPickerView.AppLabelInfo("", label, ""));
        mDataSetFiltered.clear();
        mDataSetFiltered.addAll(mDataSet);

        refreshSectionMap();
        notifyItemInserted(position);
    }

    void addSeparator(int position) {
        mDataSet.add(position,
                new AppPickerView.AppLabelInfo(AppPickerView.KEY_APP_SEPARATOR,
                        "", "").setSeparator(true));
        mDataSetFiltered.clear();
        mDataSetFiltered.addAll(mDataSet);

        refreshSectionMap();
        notifyItemInserted(position);
    }

    public void setOrder(int order) {
        mOrder = order;
        if (Build.VERSION.SDK_INT >= 24) {
            if (getAppLabelComparator(order) != null) {
                mDataSet.sort(getAppLabelComparator(order));
                mDataSetFiltered.sort(getAppLabelComparator(order));
            }
        }
        refreshSectionMap();
        notifyDataSetChanged();
    }

    private Comparator<AppPickerView.AppLabelInfo> getAppLabelComparator(int order) {
        return switch (order) {
            case AppPickerView.ORDER_ASCENDING -> APP_LABEL_ASCENDING;
            case AppPickerView.ORDER_ASCENDING_IGNORE_CASE -> APP_LABEL_ASCENDING_IGNORE_CASE;
            case AppPickerView.ORDER_DESCENDING -> APP_LABEL_DESCENDING;
            case AppPickerView.ORDER_DESCENDING_IGNORE_CASE -> APP_LABEL_DESCENDING_IGNORE_CASE;
            default -> null;
        };
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        final String packageName = mDataSetFiltered.get(position).getPackageName();
        final String activityName = mDataSetFiltered.get(position).getActivityName();

        AppPickerView.ViewHolder vh = (AppPickerView.ViewHolder) holder;
        if (!(holder instanceof AppPickerView.HeaderViewHolder)
                && !(holder instanceof AppPickerView.SeparatorViewHolder)) {
            if (mAppPickerIconLoader != null) {
                mAppPickerIconLoader.loadIcon(packageName, activityName, vh.getAppIcon());
            }

            final String label = mDataSetFiltered.get(position).getLabel();
            if (!mSearchText.isEmpty()) {
                SpannableString spannableString = new SpannableString(label);
                StringTokenizer stringTokenizer = new StringTokenizer(mSearchText);

                while (stringTokenizer.hasMoreTokens()) {
                    String nextToken = stringTokenizer.nextToken();
                    int offset = 0;
                    String remainingLabel = label;
                    do {
                        char[] prefixCharsForSpan = SeslTextUtilsReflector
                                .semGetPrefixCharForSpan(vh.getAppLabel().getPaint(), remainingLabel, nextToken.toCharArray());

                        if (prefixCharsForSpan != null) {
                            nextToken = new String(prefixCharsForSpan);
                        }

                        String lowerCaseLabel = remainingLabel.toLowerCase();
                        int tokenIndex;

                        if (remainingLabel.length() == lowerCaseLabel.length()) {
                            tokenIndex = lowerCaseLabel.indexOf(nextToken.toLowerCase());
                        } else {
                            tokenIndex = remainingLabel.indexOf(nextToken);
                        }

                        int length = nextToken.length() + tokenIndex;
                        if (tokenIndex < 0) {
                            break;
                        }

                        int tokenEndIndex = tokenIndex + offset;
                        offset += length;
                        spannableString.setSpan(
                                new ForegroundColorSpan(this.mForegroundColor), tokenEndIndex, offset, 17);
                        remainingLabel = remainingLabel.substring(length);
                        if (remainingLabel.toLowerCase().contains(nextToken.toLowerCase())) {
                            break;
                        }
                    } while (offset < MAX_OFFSET);
                }
                vh.getAppLabel().setText(spannableString);
                vh.getItem().setContentDescription(spannableString);
            } else {
                vh.getAppLabel().setText(label);
                vh.getItem().setContentDescription(label);
            }
        }

        onBindViewHolderAction(vh, position, packageName);

        if (mOnBindListener != null) {
            mOnBindListener.onBindViewHolder(vh, position, packageName);
        }
    }

    abstract void onBindViewHolderAction(AppPickerView.ViewHolder holder, int position,
            String packageName);

    @Override
    public int getItemCount() {
        return mDataSetFiltered.size();
    }

    @NonNull
    public AppPickerView.AppLabelInfo getAppInfo(int position) {
        return mDataSetFiltered.get(position);
    }

    public void setOnBindListener(@Nullable AppPickerView.OnBindListener listener) {
        mOnBindListener = listener;
    }

    public void setOnSearchFilterListener(@Nullable AppPickerView.OnSearchFilterListener listener) {
        mOnSearchFilterListener = listener;
    }

    protected boolean hasAllAppsInList() {
        return (mType == AppPickerView.TYPE_LIST_CHECKBOX_WITH_ALL_APPS
                || mType == AppPickerView.TYPE_LIST_SWITCH_WITH_ALL_APPS)
                && !mHideAllApps;
    }

    @Override
    public long getItemId(int position) {
        return mDataSetFiltered.get(position).hashCode();
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                final String searchText = constraint.toString();
                Filter.FilterResults results = new Filter.FilterResults();

                if (searchText.isEmpty()) {
                    mSearchText = "";
                    results.values = mDataSet;
                } else {
                    mSearchText = searchText;

                    ArrayList<AppPickerView.AppLabelInfo> dataSetFiltered = new ArrayList<>();
                    for (AppPickerView.AppLabelInfo labelInfo : mDataSet) {
                        if (!AppPickerView.ALL_APPS_STRING.equals(labelInfo.getPackageName())) {
                            final String label = labelInfo.getLabel();
                            if (!TextUtils.isEmpty(label)) {
                                StringTokenizer tokenizer = new StringTokenizer(searchText.toLowerCase());
                                boolean showItem = true;
                                String lowerCase = label.toLowerCase();

                                while (tokenizer.hasMoreTokens()) {
                                    if (!lowerCase.contains(tokenizer.nextToken())) {
                                        showItem = false;
                                        break;
                                    }
                                }

                                if (showItem) {
                                    dataSetFiltered.add(labelInfo);
                                }
                            }
                        }
                    }

                    results.values = dataSetFiltered;
                }
                return results;
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                mHideAllApps = !"".equals(mSearchText);
                mDataSetFiltered.clear();
                //noinspection unchecked
                mDataSetFiltered.addAll((ArrayList<AppPickerView.AppLabelInfo>) results.values);

                refreshSectionMap();
                notifyDataSetChanged();

                if (mOnSearchFilterListener != null) {
                    mOnSearchFilterListener.onSearchFilterCompleted(getItemCount());
                }
            }
        };
    }

    @Override
    public int getPositionForSection(int sectionIndex) {
        if (sectionIndex >= mSections.length) {
            return 0;
        }
        return mSectionMap.get(mSections[sectionIndex]);
    }

    @Override
    public int getSectionForPosition(int position) {
        if (position >= mPositionToSectionIndex.length) {
            return 0;
        }
        return mPositionToSectionIndex[position];
    }

    @Override
    public Object[] getSections() {
        return mSections;
    }

    void refreshSectionMap() {
        mSectionMap.clear();
        ArrayList<String> sections = new ArrayList<>();
        if (Build.VERSION.SDK_INT >= 24) {
            AlphabeticIndex.ImmutableIndex<Integer> immutableIndex = getBuckets();

            mPositionToSectionIndex = new int[mDataSetFiltered.size()];

            for (int i = 0; i < mDataSetFiltered.size(); i++) {
                String label = mDataSetFiltered.get(i).getLabel();
                if (TextUtils.isEmpty(label)) {
                    label = "";
                }
                label = immutableIndex.getBucket(immutableIndex.getBucketIndex(label)).getLabel();
                if (!mSectionMap.containsKey(label)) {
                    sections.add(label);
                    mSectionMap.put(label, i);
                }
                mPositionToSectionIndex[i] = sections.size() - 1;
            }

            mSections = new String[sections.size()];
            sections.toArray(mSections);
        }
    }

    @RequiresApi(api = 24)
    private AlphabeticIndex.ImmutableIndex<Integer> getBuckets() {
        LocaleList locales = mContext.getResources().getConfiguration().getLocales();
        if (locales.isEmpty()) {
            locales = new LocaleList(Locale.ENGLISH);
        }

        AlphabeticIndex<Integer> alphabeticIndex = new AlphabeticIndex<>(locales.get(0));
        for (int i = 1; i < locales.size(); i++) {
            alphabeticIndex.addLabels(locales.get(i));
        }
        alphabeticIndex.addLabels(Locale.ENGLISH);

        AlphabeticIndex.ImmutableIndex<Integer> immutableIndex =
                alphabeticIndex.buildImmutableIndex();
        return immutableIndex;
    }

    protected float limitFontScale(@NonNull TextView textView) {
        final float currentFontScale
                = textView.getResources().getConfiguration().fontScale;
        int i = (Float.compare(currentFontScale, 1.3f));
        final float textSize = textView.getTextSize();
        return i > 0 ? (textSize / currentFontScale) * 1.3f : textSize;
    }

    protected void limitFontLarge(@NonNull TextView textView) {
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, limitFontScale(textView));
    }

    protected void limitFontLarge2LinesHeight(@NonNull TextView textView) {
        float limitFontScale = limitFontScale(textView);
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, limitFontScale);
        textView.setMinHeight(Math.round((limitFontScale * 2.0f) + 0.5f));
    }

    private static final Comparator<AppPickerView.AppLabelInfo> APP_LABEL_ASCENDING
            = (a, b) -> {
                Collator collator = Collator.getInstance(Locale.getDefault());
                collator.setStrength(Collator.TERTIARY);
                return collator.compare(a.getLabel(), b.getLabel());
            };

    private static final Comparator<AppPickerView.AppLabelInfo> APP_LABEL_ASCENDING_IGNORE_CASE
            = (a, b) -> {
                Collator collator = Collator.getInstance(Locale.getDefault());
                collator.setStrength(Collator.PRIMARY);
                return collator.compare(a.getLabel(), b.getLabel());
            };

    private static final Comparator<AppPickerView.AppLabelInfo> APP_LABEL_DESCENDING
            = (a, b) -> {
                Collator collator = Collator.getInstance(Locale.getDefault());
                collator.setStrength(Collator.TERTIARY);
                return collator.compare(b.getLabel(), a.getLabel());
            };

    private static final Comparator<AppPickerView.AppLabelInfo> APP_LABEL_DESCENDING_IGNORE_CASE
            = (a, b) -> {
                Collator collator = Collator.getInstance(Locale.getDefault());
                collator.setStrength(Collator.PRIMARY);
                return collator.compare(b.getLabel(), a.getLabel());
            };
}

File: picker/picker-app/src/main/java/androidx/picker/adapter/AbsAdapter.kt
/*
 * Copyright 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.picker.adapter

import android.content.Context
import android.graphics.drawable.Drawable
import android.icu.text.AlphabeticIndex
import android.os.Build
import android.os.LocaleList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.SectionIndexer
import androidx.annotation.RestrictTo
import androidx.picker.R
import androidx.picker.adapter.viewholder.PickerViewHolder
import androidx.picker.common.log.LogTag
import androidx.picker.common.log.info
import androidx.picker.features.observable.UpdateMutableState
import androidx.picker.features.search.InitialSearchUtils.getMatchedStringOffset
import androidx.picker.features.search.InitialSearchUtils.getSearchResultFromSCS
import androidx.picker.loader.AppIconFlow
import androidx.picker.model.AppData
import androidx.picker.model.AppInfo
import androidx.picker.model.AppInfoData
import androidx.picker.model.AppInfoDataImpl
import androidx.picker.model.GroupTitleStyleData
import androidx.picker.model.Highlightable
import androidx.picker.model.viewdata.AppInfoViewData
import androidx.picker.model.viewdata.AppSideViewData
import androidx.picker.model.viewdata.CategoryViewData
import androidx.picker.model.viewdata.GroupTitleViewData
import androidx.picker.model.viewdata.SearchableViewData
import androidx.picker.model.viewdata.ViewData
import androidx.picker.widget.SeslAppPickerView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import java.util.Locale
import java.util.StringTokenizer
import kotlin.collections.ArrayList
import kotlin.collections.List
import kotlin.collections.set
import kotlin.reflect.KProperty
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.FlowCollector

/**
 * Abstract base class for RecyclerView adapters used in the [SeslAppPickerView].
 * This class provides common functionality for handling data, filtering, and section indexing.
 *
 * @param context The context used for accessing resources and other system services.
 * @param groupTitleStyleData The style data for group titles. Defaults to [GroupTitleStyleData.SOLID].
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP_PREFIX)
abstract class AbsAdapter @JvmOverloads constructor(
    protected val context: Context,
    protected val groupTitleStyleData: GroupTitleStyleData = GroupTitleStyleData.SOLID
) : RecyclerView.Adapter<PickerViewHolder>(), AppPickerAdapter, SectionIndexer, LogTag {

    private val dataListUnfiltered: MutableList<ViewData> = ArrayList()
    private val dataListFiltered: MutableList<ViewData> = ArrayList()
    private var _filter: Filter? = null
    private var searchText: String = ""

    private var onBindListener: AppPickerAdapter.OnBindListener? = null
    private var onSearchFilterListener: SeslAppPickerView.OnSearchFilterListener? = null
    private var positionToSectionIndex = intArrayOf()
    private val sectionMap = HashMap<String, Int>()
    private var _sections = arrayOf<Any>()

    private fun convertCategoryViewData2AppInfoViewData(categoryViewData: CategoryViewData): AppInfoViewData {
        val checkBoxAppInfoData = AppData.ListCheckBoxAppDataBuilder(categoryViewData.appData.appInfo)
            .setLabel(categoryViewData.title)
            .setIcon(categoryViewData.icon)
            .setSelected(categoryViewData.selectableItem.isSelected)
            .build() as AppInfoDataImpl

        return AppInfoViewData(
            appInfoData = checkBoxAppInfoData,
            iconFlow = AppIconFlow(
                base = object : UpdateMutableState<AppInfoData, Drawable>(checkBoxAppInfoData) {
                    override fun setValue(thisRef: Any?, prop: KProperty<*>, value: Drawable?) {
                        checkBoxAppInfoData.icon = value
                    }
                    override fun getValue(thisRef: Any?, prop: KProperty<*>): Drawable? =
                        checkBoxAppInfoData.icon
                },
                defaultIconFlow = object : Flow<Drawable?> {
                    override suspend fun collect(collector: FlowCollector<Drawable?>) {
                        collector.emit(categoryViewData.icon)
                    }
                }
            ),
            selectableItem = categoryViewData.selectableItem,
            spanCount = -1,
            onActionClick = null
        )
    }

    private fun isFilterMatch(searchText: String, searchable: String?): Boolean {
        if (searchable.isNullOrEmpty()) return false
        val tokens = StringTokenizer(searchText.lowercase())
        val searchableLower = searchable.lowercase().trim().replace(" ", "")
        val pattern = this@AbsAdapter.searchText.trim().replace(" ", "")
        while (tokens.hasMoreTokens()) {
            val token = tokens.nextToken()
            if (searchableLower.contains(token) || getMatchedStringOffset(searchableLower, pattern) > -1) return true
        }
        return false
    }

    private fun refreshSectionMap(list: List<ViewData>) {
        sectionMap.clear()
        positionToSectionIndex = IntArray(list.size)
        val sectionsLocal = mutableListOf<String>()

        if (Build.VERSION.SDK_INT >= 24) {
            var locales = context.resources.configuration.locales.takeIf { !it.isEmpty }
                ?: LocaleList(Locale.ENGLISH)
            val alphabeticIndex = AlphabeticIndex<Locale>(locales[0])

            for (i in 1 until locales.size()) {
                alphabeticIndex.addLabels(locales[i])
            }

            alphabeticIndex.addLabels(Locale.ENGLISH)
            val buildImmutableIndex = alphabeticIndex.buildImmutableIndex()

            for (i in list.indices) {
                val viewData = list[i]
                if (viewData is AppInfoViewData) {
                    var label = viewData.label ?: ""
                    val sectionLabel = buildImmutableIndex.getBucket(buildImmutableIndex.getBucketIndex(label)).label
                    if (!sectionMap.containsKey(sectionLabel)) {
                        sectionsLocal.add(sectionLabel)
                        sectionMap[sectionLabel] = i
                    }
                    positionToSectionIndex[i] = sectionsLocal.size - 1
                }
            }
        } else {
            for (i in list.indices) {
                val viewData = list[i]
                if (viewData is AppInfoViewData) {
                    var label = viewData.label ?: ""
                    val sectionLabel = label.first().toString()
                    if (!sectionMap.containsKey(label)) {
                        sectionsLocal.add(sectionLabel)
                        sectionMap[sectionLabel] = i
                    }
                    positionToSectionIndex[i] = sectionsLocal.size - 1
                }
            }
        }
        _sections = sectionsLocal.toTypedArray()
    }

    private fun replace(list: MutableList<ViewData>, old: ViewData, new: ViewData) {
        val index = list.indexOf(old)
        if (index == -1) return
        list.removeAt(index)
        list.add(index, new)
    }

    /**
     * Appends a list of [ViewData] to the current dataset and updates the adapter.
     *
     * @param list The list of [ViewData] to append.
     */
    fun appendItemList(list: List<ViewData>) {
        val arrayList = ArrayList(dataListUnfiltered)
        arrayList.addAll(list)
        submitList(arrayList)
    }

    /**
     * Processes a list of [ViewData] to handle invisible children of [CategoryViewData].
     *
     * This function iterates through the input `viewDataList`.
     * - If a [CategoryViewData] is encountered, its `invisibleChildren` are added to a temporary list.
     * - If an [AppInfoViewData] is encountered and it's present in the temporary list of
     *   invisible children, its `highlightText` is set to the current `searchText`.
     *
     * This is used to ensure that even if an app is an invisible child of a category (meaning it
     * might not be directly displayed under that category due to filtering or other logic),
     * it still gets highlighted if it matches the search query.
     *
     * @param viewDataList The list of [ViewData] to process.
     * @return A new list of [ViewData] with highlighting applied to relevant invisible children.
     */
    fun applyInvisibleChildrenOfCategoryAppData(viewDataList: List<ViewData>): List<ViewData> {
        val processedViewDataList = ArrayList<ViewData>()
        val invisibleChildrenFromCategories = ArrayList<ViewData>()
        for (viewData in viewDataList) {
            if (viewData is CategoryViewData) {
                invisibleChildrenFromCategories.addAll(viewData.invisibleChildren)
            } else if (viewData is AppInfoViewData && invisibleChildrenFromCategories.contains(
                    viewData
                )
            ) {
                viewData.getHighlightText().setValue(searchText)
            }
            processedViewDataList.add(viewData)
        }
        return processedViewDataList
    }

    /**
     * Converts a list of [CategoryViewData] objects into a list of [ViewData] objects,
     * specifically by transforming each [CategoryViewData] into an [AppInfoViewData].
     * This is typically used when displaying categories as app-like items in a filtered list.
     *
     * @param list The list of [CategoryViewData] to be converted.
     * @return A new list of [ViewData] containing the converted [AppInfoViewData] objects.
     */
    fun generateCategoryFilterResult(list: List<CategoryViewData>): List<ViewData> {
        val arrayList = ArrayList<ViewData>()
        for (cat in list) {
            arrayList.add(convertCategoryViewData2AppInfoViewData(cat))
        }
        return arrayList
    }

    /**
     * Generates a header for a filtered list of app side view data.
     *
     * @param str The title string for the header.
     * @param list The list of AppSideViewData items to be included under this header.
     * @return A ViewData object representing the group title header.
     */
    fun generateFilterHeader(str: String, list: List<AppSideViewData>): ViewData {
        val map = list.map { sideViewData -> sideViewData.appData }
        return GroupTitleViewData(
            AppData.GroupAppDataBuilder(str)
                .setLabel(str)
                .setSubLabel(map.size.toString())
                .setAppDatas(map)
                .build()
        )
    }

    override fun getAppInfo(position: Int): ViewData = dataListFiltered[position]

    /**
     * Filters a list of [SearchableViewData] based on the current [searchText] and a provided list of [AppInfo].
     *
     * This function iterates through each [SearchableViewData] in the [searchableViewDataList] list.
     * For each item, it checks if any of its `searchable` terms match the [searchText] using the [isFilterMatch] function.
     * If no direct match is found and the `key` of the [SearchableViewData] is an [AppInfo] instance,
     * it then checks if this [AppInfo] is present in the [appInfoList].
     * If either of these conditions is true, the [SearchableViewData] item is added to the result list.
     *
     * @param T The type of [SearchableViewData] being filtered.
     * @param searchableViewDataList The list of [SearchableViewData] items to filter.
     * @param appInfoList A list of [AppInfo] objects to check against if a direct search term match is not found.
     * @return A new list containing only the [SearchableViewData] items that matched the filter criteria.
     */
    fun <T : SearchableViewData> getAppInfoFilterResult(
        searchableViewDataList: List<T>,
        appInfoList: List<AppInfo>
    ): List<T> {
        val arrayList = ArrayList<T>()
        for (searchableData in searchableViewDataList) {
            var isMatch = false
            for (searchable in searchableData.searchable) {
                if (isFilterMatch(searchText, searchable)) {
                    isMatch = true
                    break
                }
            }
            if (!isMatch && searchableData.key is AppInfo) {
                isMatch = appInfoList.contains(searchableData.key as AppInfo)
            }
            if (isMatch) {
                arrayList.add(searchableData)
            }
        }
        return arrayList
    }

    override fun getDataSetFiltered(): MutableList<ViewData> = dataListFiltered

    override fun getFilter(): Filter {
        if (_filter != null) return _filter!!
        _filter = object : Filter() {
            override fun performFiltering(currentSearchText: CharSequence?): FilterResults {
                searchText = currentSearchText?.toString() ?: ""
                val results = FilterResults()

                val originalList = dataListUnfiltered
                val filteredList = mutableListOf<ViewData>()

                if (searchText.isEmpty()) {
                    filteredList.addAll(applyInvisibleChildrenOfCategoryAppData(originalList))
                } else {
                    val resultsFromSCS = getSearchResultFromSCS(context, searchText)
                    val catViewDataList = mutableListOf<CategoryViewData>()
                    val appInfoViewDataList = mutableListOf<AppInfoViewData>()

                    originalList.forEach {
                        if (it is CategoryViewData) {
                            catViewDataList.add(it)
                        } else if (it is AppInfoViewData) {
                            appInfoViewDataList.add(it)
                        }
                    }

                    val catViewDataListFiltered = getAppInfoFilterResult(catViewDataList, resultsFromSCS)
                    if (catViewDataListFiltered.isNotEmpty()) {
                        filteredList.add(generateFilterHeader(categoriesTitle, catViewDataListFiltered))
                        filteredList.addAll(generateCategoryFilterResult(catViewDataListFiltered))
                    }

                    val appInfoViewDataListFiltered = getAppInfoFilterResult(appInfoViewDataList, resultsFromSCS)
                    if (appInfoViewDataListFiltered.isNotEmpty()) {
                        if (catViewDataList.isNotEmpty()) {
                            filteredList.add(generateFilterHeader(appTitle, appInfoViewDataListFiltered))
                        }
                        filteredList.addAll(appInfoViewDataListFiltered)
                    }
                }
                results.values = filteredList
                return results
            }

            override fun publishResults(
                charSequence: CharSequence?,
                filterResults: FilterResults?
            ) {
                @Suppress("UNCHECKED_CAST")
                val results = filterResults?.values as? ArrayList<ViewData> ?: return
                for (viewData in results) {
                    if (viewData is Highlightable) {
                        viewData.getHighlightText().setValue(searchText)
                    }
                }
                onUpdateFilteredList(results)
                onSearchFilterListener?.onSearchFilterCompleted(itemCount)
            }
        }
        return _filter!!
    }

    override fun getItemCount(): Int = dataListFiltered.size

    override fun getItemId(i: Int): Long = dataListFiltered[i].key.hashCode().toLong()

    override fun getPositionForSection(i: Int): Int {
        val sections = _sections
        if (i < sections.size) {
            val position = sectionMap[sections[i]]
            if (position != null) return position
        }
        return 0
    }

    override fun getSectionForPosition(position: Int): Int {
        val positionToSectionIndex = positionToSectionIndex
        if (position >= positionToSectionIndex.size) return 0
        return positionToSectionIndex[position]
    }

    override fun getSections(): Array<Any> = _sections

    fun inflate(viewGroup: ViewGroup, layoutRes: Int): View {
        return LayoutInflater.from(viewGroup.context).inflate(layoutRes, viewGroup, false)
    }

    fun onUpdateFilteredList(list: List<ViewData>) {
        val calculateDiff = DiffUtil.calculateDiff(DiffUtilCallback(dataListFiltered, list))
        dataListFiltered.clear()
        dataListFiltered.addAll(list)
        refreshSectionMap(list)
        calculateDiff.dispatchUpdatesTo(NearbyListUpdateCallback(this))
    }

    override fun setOnBindListener(listener: AppPickerAdapter.OnBindListener) {
        onBindListener = listener
    }

    override fun setOnSearchFilterListener(listener: SeslAppPickerView.OnSearchFilterListener) {
        onSearchFilterListener = listener
    }

    override fun submitList(itemList: List<ViewData>) {
        info("submitList list=${itemList.size}")
        dataListUnfiltered.clear()
        dataListUnfiltered.addAll(itemList)
        getFilter().filter(searchText)
    }

    override fun updateItem(viewData: ViewData) {
        val arrayList = ArrayList(dataListUnfiltered)
        for (next in dataListUnfiltered) {
            if (next.key == viewData.key) {
                replace(arrayList, next, viewData)
                break
            }
        }
        submitList(arrayList)
    }

    override fun onBindViewHolder(holder: PickerViewHolder, position: Int) {
        val viewData = dataListFiltered[position]
        onBindListener?.onBindViewHolder(holder, viewData)
        holder.bindData(viewData)
        holder.bindAdapter(this)
    }

    override fun onBindViewHolder(holder: PickerViewHolder, position: Int, payloads: List<Any>) {
        if (payloads.isEmpty()) {
            super.onBindViewHolder(holder, position, payloads)
        } else {
            onBindViewHolder(holder, position)
        }
    }

    internal inline val categoriesTitle get() = context.resources.getString(R.string.title_categories)
    internal inline val appTitle get() = context.resources.getString(R.string.title_apps)

    override val logTag: String = "AppPickerViewAdapter"
}